lms: Move merkle tree generation to heap allocation

mfischer opened the pull request:
Larger height (e.g. H=20) trees cannot be put on the stack. Allocate memory for them based on need using mbedtls_calloc().

Signed-off-by: Moritz Fischer <moritzf@google.com>

## Description

This is in preparation of adding H=20 support in a separate PR https://github.com/Mbed-TLS/mbedtls/pull/6584.


## Gatekeeper checklist

- [ ] **changelog** provided, or not required
- [ ] **backport** done, or not required
- [ ] **tests** provided, or not required



## Notes for the submitter

Please refer to the [contributing guidelines](../CONTRIBUTING.md), especially the
checklist for PR contributors.

mfischer commented:
@RcColes @gilles-peskine-arm

gilles-peskine-arm commented on library/lms.c:
```diff
@@ -65,7 +65,6 @@
 /* Currently only support H=10 */
 #define H_TREE_HEIGHT_MAX                  10
-#define MERKLE_TREE_NODE_AM_MAX            (1u << (H_TREE_HEIGHT_MAX + 1u))
 #define MERKLE_TREE_NODE_AM(type)          (1u << (MBEDTLS_LMS_H_TREE_HEIGHT(type) + 1u))
```
MSVC is unhappy, see https://github.com/Mbed-TLS/mbedtls/pull/6584#discussion_r1023147349

mfischer replied:
Reading MSVC suggestion on their website they want the whole thing to be 64 bit in that case I think.

Took a stab. Let's see if that silences it, otherwise will try the size_t cast.

Thanks for your help!

mfischer replied:
This looks different, any chance this is unrelated?

gilles-peskine-arm replied:
It was the same error message, and your fix made it pass, so all good here.

gilles-peskine-arm commented on library/lms.c:
```diff
@@ -499,13 +498,20 @@ static int get_merkle_path( mbedtls_lms_private_t *ctx,
 
-    ret = calculate_merkle_tree( ctx, ( unsigned char * )tree );
+    tree = mbedtls_calloc( MERKLE_TREE_NODE_AM(ctx->params.type),
+        MBEDTLS_LMS_M_NODE_BYTES(ctx->params.type) );
```
Partly preexisting: defining a short name (`const size_t …`) for `MBEDTLS_LMS_M_NODE_BYTES( ctx->params.type )`, and perhaps also `MERKLE_TREE_NODE_AM( ctx->params.type )`, would make the code easier to read.

Also goes in `mbedtls_lms_calculate_public_key`.

mfischer replied:
Done.

gilles-peskine-arm commented on library/lms.c:
```diff
@@ -499,13 +498,20 @@ static int get_merkle_path( mbedtls_lms_private_t *ctx,
-    ret = calculate_merkle_tree( ctx, ( unsigned char * )tree );
+    tree = mbedtls_calloc( MERKLE_TREE_NODE_AM(ctx->params.type),
+        MBEDTLS_LMS_M_NODE_BYTES(ctx->params.type) );
+    if ( !tree )
```
Style: for readability, we avoid implicit truthiness of integers and pointers.
```suggestion
    if ( tree == NULL )
```

Also goes in `mbedtls_lms_calculate_public_key`.

(Note: there are violations of our current style rules around parentheses and whitespace. We're going to change those rules soon, so I'm not going to insist on the current rules.)

mfischer replied:
Done

gilles-peskine-arm reviewed (changes requested):
Looks mostly good to me, just a few minor things.

mfischer commented on library/lms.c:
```diff
@@ -65,10 +65,9 @@
-#define MERKLE_TREE_NODE_AM(type)          (1u << (MBEDTLS_LMS_H_TREE_HEIGHT(type) + 1u))
-#define MERKLE_TREE_LEAF_NODE_AM(type)     (1u << MBEDTLS_LMS_H_TREE_HEIGHT(type))
-#define MERKLE_TREE_INTERNAL_NODE_AM(type) (1u << MBEDTLS_LMS_H_TREE_HEIGHT(type))
+#define MERKLE_TREE_NODE_AM(type)          (1ull << (MBEDTLS_LMS_H_TREE_HEIGHT(type) + 1u))
```
Done

gilles-peskine-arm reviewed (approved):
Looks good to me.

For future reference, please don't rebase a pull request during review if it isn't needed. (Needed is, for example, to fix a commit message, or because there's a conflict with the base branch.) Work done in response to change requests should go into new commits (broken down into meaningful pieces, not a giant “addressed review comments” commit). Rebasing makes the reviewer's job more difficult because the differences aren't grouped, it's harder to keep track of where you left off, and it leaves an inaccurate history.

mprse commented on library/lms.c:
```diff
@@ -516,17 +523,18 @@ static int get_merkle_path( mbedtls_lms_private_t *ctx,
-                &tree[adjacent_node_id],
-                MBEDTLS_LMS_M_NODE_BYTES(ctx->params.type) );
+        memcpy( &path[height * node_bytes],
+                &tree[adjacent_node_id * node_bytes], node_bytes );
```
Why the second parameter (source address) has been changed?
`&tree[adjacent_node_id]` -> `&tree[adjacent_node_id * node_bytes]`

mfischer replied:
The type of tree changed, now it's a flat unsigned char array.

mprse commented on library/lms.c:
```diff
@@ -499,13 +498,21 @@ static int get_merkle_path( mbedtls_lms_private_t *ctx,
                             unsigned char *path )
 {
-    unsigned char tree[MERKLE_TREE_NODE_AM_MAX][MBEDTLS_LMS_M_NODE_BYTES_MAX];
+    const size_t node_bytes = MBEDTLS_LMS_M_NODE_BYTES(ctx->params.type);
```
```suggestion
    const size_t node_bytes = MBEDTLS_LMS_M_NODE_BYTES( ctx->params.type );
```

gilles-peskine-arm replied:
We're [switching to the new style](https://github.com/Mbed-TLS/mbedtls/issues/6359) at the end of the month (or maybe first week of January). So let's not be picky about it now.

mprse commented on library/lms.c:
```diff
@@ -499,13 +498,21 @@ static int get_merkle_path( mbedtls_lms_private_t *ctx,
     int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
 
-    ret = calculate_merkle_tree( ctx, ( unsigned char * )tree );
+    tree = mbedtls_calloc( MERKLE_TREE_NODE_AM(ctx->params.type),
```
```suggestion
    tree = mbedtls_calloc( MERKLE_TREE_NODE_AM( ctx->params.type ),
```

mprse commented on library/lms.c:
```diff
@@ -499,13 +498,21 @@ static int get_merkle_path( mbedtls_lms_private_t *ctx,
-    ret = calculate_merkle_tree( ctx, ( unsigned char * )tree );
+    tree = mbedtls_calloc( MERKLE_TREE_NODE_AM(ctx->params.type),
+        node_bytes );
+    if ( tree == NULL )
```
```suggestion
    if( tree == NULL )
        return MBEDTLS_ERR_LMS_ALLOC_FAILED;
```

mprse commented on library/lms.c:
```diff
@@ -516,17 +523,18 @@ static int get_merkle_path( mbedtls_lms_private_t *ctx,
 exit:
-    mbedtls_platform_zeroize( tree, sizeof( tree ) );
+    mbedtls_platform_zeroize( tree, node_bytes *
+        MERKLE_TREE_NODE_AM(ctx->params.type) );
```
```suggestion
        MERKLE_TREE_NODE_AM( ctx->params.type ) );
```

mprse commented on library/lms.c:
```diff
@@ -516,17 +523,18 @@ static int get_merkle_path( mbedtls_lms_private_t *ctx,
-    mbedtls_platform_zeroize( tree, sizeof( tree ) );
+    mbedtls_platform_zeroize( tree, node_bytes *
+        MERKLE_TREE_NODE_AM(ctx->params.type) );
+    mbedtls_free ( tree );
```
```suggestion
    mbedtls_free( tree );
```

mprse commented on library/lms.c:
```diff
@@ -659,8 +667,9 @@ int mbedtls_lms_generate_private_key( mbedtls_lms_private_t *ctx,
                                       const mbedtls_lms_private_t *priv_ctx )
 {
-    unsigned char tree[MERKLE_TREE_NODE_AM_MAX][MBEDTLS_LMS_M_NODE_BYTES_MAX];
+    const size_t node_bytes = MBEDTLS_LMS_M_NODE_BYTES(priv_ctx->params.type);
```
```suggestion
    const size_t node_bytes = MBEDTLS_LMS_M_NODE_BYTES( priv_ctx->params.type );
```

mprse commented on library/lms.c:
```diff
@@ -679,25 +688,33 @@ int mbedtls_lms_calculate_public_key( mbedtls_lms_public_t *ctx,
         return( MBEDTLS_ERR_LMS_BAD_INPUT_DATA );
     }
 
+    tree = mbedtls_calloc( MERKLE_TREE_NODE_AM(priv_ctx->params.type),
```
```suggestion
    tree = mbedtls_calloc( MERKLE_TREE_NODE_AM( priv_ctx->params.type ),
```

mprse commented on library/lms.c:
```diff
@@ -679,25 +688,33 @@ int mbedtls_lms_calculate_public_key( mbedtls_lms_public_t *ctx,
 
+    tree = mbedtls_calloc( MERKLE_TREE_NODE_AM(priv_ctx->params.type),
+        node_bytes );
+    if ( tree == NULL )
```
```suggestion
    if ( tree == NULL )
        return MBEDTLS_ERR_LMS_ALLOC_FAILED;
```

mprse commented on library/lms.c:
```diff
@@ -679,25 +688,33 @@ int mbedtls_lms_calculate_public_key( mbedtls_lms_public_t *ctx,
     /* Root node is always at position 1, due to 1-based indexing */
-    memcpy( ctx->T_1_pub_key, &tree[1],
-            MBEDTLS_LMS_M_NODE_BYTES(ctx->params.type) );
+    memcpy( ctx->T_1_pub_key, &tree[node_bytes], node_bytes );
```
Why the second parameter (source address) has been changed?
`&tree[1]` -> `&tree[node_bytes]`

The comment above informs why index `1` was used here which is not the case now.

mfischer replied:
Type of tree changed to a flat unsigned char array. Any suggestions on how to reword the comment?

mprse commented on library/lms.c:
```diff
@@ -679,25 +688,33 @@ int mbedtls_lms_calculate_public_key( mbedtls_lms_public_t *ctx,
 exit:
-    mbedtls_platform_zeroize( tree, sizeof( tree ) );
+    mbedtls_platform_zeroize( tree, node_bytes *
+        MERKLE_TREE_NODE_AM(priv_ctx->params.type) );
```
```suggestion
        MERKLE_TREE_NODE_AM( priv_ctx->params.type ) );
```

mprse commented on library/lms.c:
```diff
@@ -679,25 +688,33 @@ int mbedtls_lms_calculate_public_key( mbedtls_lms_public_t *ctx,
-    mbedtls_platform_zeroize( tree, sizeof( tree ) );
+    mbedtls_platform_zeroize( tree, node_bytes *
+        MERKLE_TREE_NODE_AM(priv_ctx->params.type) );
+    mbedtls_free ( tree );
```
```suggestion
    mbedtls_free( tree );
```

mprse reviewed (commented):
Generally looks good.
Marked only style issues (I don't know if they are important now as we will change code style soon anyway) and left two questions to clarify.

mfischer commented:
> Looks good to me.
> 
> For future reference, please don't rebase a pull request during review if it isn't needed. (Needed is, for example, to fix a commit message, or because there's a conflict with the base branch.) Work done in response to change requests should go into new commits (broken down into meaningful pieces, not a giant “addressed review comments” commit). Rebasing makes the reviewer's job more difficult because the differences aren't grouped, it's harder to keep track of where you left off, and it leaves an inaccurate history.

Thanks for the explanation. In terms of process will you guys then squash the commits together when merging?

I'm not super familiar with github flows for this (yet, I'm used to sending patches by email to mailing list).

gilles-peskine-arm commented:
@mfischer We don't squash. We want to retain the development history in version control, not just the merge history. It's very useful when you want to understand how the code came to be the way it is.

mprse reviewed (approved):
LGTM.
Style issues ignored.
